pkg/logs/sender/sender.go

```diff
@@ -96,3 +109,70 @@ func (s *Sender) send(payload []byte) error {
 func shouldStopSending(err error) bool {
 	return err == context.Canceled
 }
+
+// SplitSenders splits a single stream of message into 2 equal streams.
+// Acts like an AND gate in that the input will only block if both outputs block.
+// This ensures backpressure is propagated to the input to prevent loss of measages in the pipeline.
+func SplitSenders(inputChan chan *message.Message, main *Sender, backup *Sender) {
+	go func() {
+		mainSenderHasErr := false
+		backupSenderHasErr := false
+
+		for message := range inputChan {
+			sentMain := false
+			sentBackup := false
+
+			// First collect any errors from the senders
+			select {
+			case mainSenderHasErr = <-main.hasError:
+			default:
+			}
+
+			select {
+			case backupSenderHasErr = <-backup.hasError:
+			default:
+			}
+
+			// If both senders are failing, we want to block the pipeline until at least one succeeds
+			if mainSenderHasErr && backupSenderHasErr {
+				select {
+				case main.inputChan <- message:
+					sentMain = true
+				case backup.inputChan <- message:
+					sentBackup = true
+				case mainSenderHasErr = <-main.hasError:
+				case backupSenderHasErr = <-backup.hasError:
+				}
+			}
+
+			if !sentMain {
+				mainSenderHasErr = sendMessage(mainSenderHasErr, main, message)
+			}
+
+			if !sentBackup {
+				backupSenderHasErr = sendMessage(backupSenderHasErr, backup, message)
+			}
+		}
+	}()
+}
+
+func sendMessage(hasError bool, sender *Sender, message *message.Message) bool {
+	if !hasError {
+		// If there is no error - block and write to the buffered channel until it succeeds or we get an error.
+		// If we don't block, the input can fill the buffered channels faster than sender can
+		// drain them - causing missing logs.
+		select {
+		case sender.inputChan <- message:
+		case hasError = <-sender.hasError:
+		}
+	} else {
+		// Even if there is an error, try to put the log line in the buffered channel in case the
+		// error resolves quickly and there is room in the channel.
+		select {
+		case sender.inputChan <- message:
+		default:
+			break
```

Comment:
nit I'm not sure what the break does here - if you want to get back to the caller I think `return nil` may be clearer.

Reply:
Good catch - it's a habit. some languages force you to put a statement in a default case (swift for example) 😆